### PR TITLE
233 revise proposal confirmation

### PR DESCRIPTION
**Pull request by mbburch**

### CODE REVIEW

File: app/decorators/staff/program_session_decorator.rb [hunk 1]
```diff
@@ -39,18 +38,36 @@ def track_name
...
       ts = object.time_slot
-      parts << ts.conference_day if ts.conference_day.present?
-      parts << ts.start_time.to_s(:time) if ts.start_time.present?
-      parts << ts.room.name if ts.room.present?
+      "Day #{ts.conference_day}" if ts.conference_day.present?
+    end
+  end
+
+  def complete_video_url
+    if object.video_url.include?("://")
```
Comment (mghaught) [inline app/decorators/staff/program_session_decorator.rb]: how would this look if you used a regex instead?

File: app/models/proposal.rb [hunk 2]
```diff
@@ -118,13 +124,15 @@ def withdraw
...
 
   def confirm
-    transaction do
-      update!(confirmed_at: DateTime.current)
-      ps = ProgramSession.create_from_proposal(self)
-      ps.persisted?
-    end
-  rescue ActiveRecord::RecordInvalid
-    false
+    self.update(confirmed_at: DateTime.current)
```
Comment (mghaught) [inline app/models/proposal.rb]: I don't believe you need the use of self on these two lines.  Try removing it and see how it works.

File: app/models/proposal.rb [hunk 3]
```diff
@@ -118,13 +124,15 @@ def withdraw
...
+    ps = self.program_session
+    ps.update(state: self.waitlisted? ? ProgramSession::WAITLISTED : ProgramSession::LIVE) if ps.present?
```
Comment (mghaught) [inline app/models/proposal.rb]: try inlining these two lines.  what I'd expect is something like this:

program_session.update(...)

I'd also find this line a bit more readable by leading with the if block such as:

```
if program_session.present?
  program_session.update(...)
end
```
I also found the ? inline was a bit opaque and would love to see that split off such as:

```
if program_session.present?
  new_state = waitlisted? ? ProgramSession::WAITLISTED : ProgramSession::LIVE
  program_session.update(state: new_state)
end
```

File: app/models/proposal.rb [hunk 4]
```diff
@@ -118,13 +124,15 @@ def withdraw
...
+  end
+
+  def decline
+    self.update(state: WITHDRAWN)
```
Comment (mghaught) [inline app/models/proposal.rb]: Look at removing self.  Then combine the two updates into a single call.  Right now it's doing two database hits for updating the proposal.  update can take multiple attributes so there's no reason not to combine them.

File: spec/features/proposal_spec.rb [hunk 5]
```diff
@@ -246,4 +249,31 @@
       expect(page).to have_text(proposal.title)
     end
   end
+
+  context "when declined" do
+    let(:proposal) { create(:proposal) }
+    let(:speaker) { create(:speaker, proposal: proposal, user: user) }
+
+    before do
+      proposal.speakers << speaker
```
Comment (mghaught) [inline spec/features/proposal_spec.rb]: I'm curious why these two lines are handled by the let blocks.

**Review by mghaught — CHANGES_REQUESTED** [review]

File: spec/features/proposal_spec.rb (hunk 5 above)
Comment (mbburch) [reply spec/features/proposal_spec.rb]: Good question. I modeled it after the other tests in this spec. Would you recommend putting it inside the before block instead? @mghaught

**Pull request merged**